examples/general/basic_dead_store/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_ast;
+extern crate rustc_hir;
+extern crate rustc_span;
+use clippy_utils::{diagnostics::span_lint_and_help, get_parent_expr};
+
+use rustc_ast::LitKind;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_span::Span;
+
+dylint_linting::impl_late_lint! {
+    /// ### What it does
+    /// Finds instances of dead stores in arrays: array positions that are assigned twice without a
+    ///  use or read in between.
+    ///
+    /// ### Why is this bad?
+    /// A dead store might indicate a logic error in the program or an unnecessary assignment.
+    ///
+    /// ### Known problems
+    /// This lint only checks for literal indices and will not try to find instances where an array
+    /// is indexed by a variable.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut arr = [0u64; 2];
+    /// arr[0] = 1;
+    /// arr[0] = 2;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut arr = [0u64; 2];
+    /// arr[0] = 2;
+    /// arr[1] = 1;
+    /// ```
+    pub BASIC_DEAD_STORE,
+    Warn,
+    "An array element is assigned twice without a use or read in between",
+    BasicDeadStore::default()
+}
+
+#[derive(Default)]
+pub struct BasicDeadStore {
+    /// Stores instances of array-indexing with literal (array name, index, span)
+    arr_and_idx_vec: Vec<(String, u128, Span)>,
+}
+
+impl BasicDeadStore {
+    /// Removes all stored values of the given array
+    fn clear_stores_of(&mut self, string: &String) {
+        self.arr_and_idx_vec
+            .retain(|(arr_string, _idx, _span)| arr_string != string);
+    }
+
+    /// Returns all instances where the given array is indexed at `idx`
+    fn get_pairs_with_same_name_idx(
+        &self,
+        string: &String,
+        idx: &u128,
+    ) -> Vec<&(String, u128, Span)> {
+        self.arr_and_idx_vec
+            .iter()
+            .filter(|(arr_string, arr_idx, _span)| arr_string == string && arr_idx == idx)
+            .collect()
+    }
+}
+
+/// Checks if the given expression is an assignment to an array indexed by a literal.
+/// Returns the tuple (array name, indexed position, span)
+fn is_assignment_to_array_indexed_by_literal(
+    expr: &Expr,
+    arr_string: &String,
+    tcx: &LateContext<'_>,
+) -> Option<(String, u128, Span)> {
+    let index_expr = get_parent_expr(tcx, expr)?;
+    if let ExprKind::Index(array, index, _span) = index_expr.kind {
+        if array.hir_id == expr.hir_id {
+            let assign_expr = get_parent_expr(tcx, index_expr)?;
+            if let ExprKind::Assign(target, _value, assignment_span) = assign_expr.kind {
+                if target.hir_id == index_expr.hir_id {
+                    if let ExprKind::Lit(lit) = index.kind {
+                        if let LitKind::Int(index, _type) = lit.node {
+                            return Some((arr_string.to_string(), index, assignment_span));
```

Comment:
We could leave the first component a `String`, but would it make more sense to return an `&Expr` and to compare them using `clippy_util`'s [`eq_expr`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/struct.SpanlessEq.html#method.eq_expr)?

This code is dated, but here is an example of `eq_expr` being used: https://github.com/crytic/solana-lints/blob/3c7044bc874ec7aeaee93e0fcf88a10793057bd4/lints/missing_owner_check/src/lib.rs#L111-L112

Reply:
I simplified the logic in the latest commit -- there was an unnecessary argument passing. Saving expr's required me to handle lifetimes on the BasicDeadStore struct and I felt that comparing the strings was fine. Do you think it is worth using eq_expr?

Reply:
Actually, there may be a better way still, using `expr_local`: https://doc.rust-lang.org/stable/nightly-rustc/clippy_utils/mir/fn.expr_local.html

But we can leave it as strings for now. That way we have a set of working tests to compare a new approach to.